spark/dl/src/main/scala/com/intel/analytics/bigdl/parameters/AllReduceParameter.scala

```diff
@@ -126,54 +141,69 @@ class AllReduceParameter[T: ClassTag](id: Long, partitionNum: Int,
     BlockManagerWrapper.putBytes(blockId, fp16param.bytes(), StorageLevel.MEMORY_ONLY_SER)
   }
 
-  def getWeightBlockId(pid : Int): BlockId = {
+  private def getWeightBlockId(pid: Int): BlockId = {
     SparkExtension.getLocalBlockId(id + "weightBytes" + pid)
   }
 
-  def getWeightPartitionId(): BlockId = {
+  private def getWeightPartitionId(): BlockId = {
     SparkExtension.getLocalBlockId(id + "weights" + partitionId)
   }
 
-  def getGradientPartitionId(): BlockId = {
+  private def getGradientPartitionId(): BlockId = {
     SparkExtension.getLocalBlockId(id + "gradients" + partitionId)
   }
 
-  def getGradientBlockId(pidFrom : Int, pidTo : Int): BlockId = {
+  private def getGradientBlockId(pidFrom: Int, pidTo: Int): BlockId = {
     SparkExtension.getLocalBlockId(id.toString + pidTo + "gradientBytes" + pidFrom)
   }
 
+  /**
+   * Use a fixed thread pool to launch a thread for each partition of the weights. Each thread
+   * requests a partition of the weights from the Spark block manager and copies it into
+   * `localParameter`.
+   *
+   * @param localParameter The Tensor that will hold the retrieved weights.
+   * @return A `FutureResult` which contains futures for each thread.
+   */
   def getWeights(localParameter: Tensor[T]): FutureResult[Int] = {
-  val bm = SparkEnv.get.blockManager
-    val tasks = (0 until partitionNum).map(pid => {
-      syncPool.submit(new Callable[Int] {
-        override def call(): Int = {
-          try {
-            val blockId = getWeightBlockId(pid)
-            val localBuffer = BlockManagerWrapper.byteBufferConvert(
-              bm.getLocalBytes(blockId).getOrElse(bm.getRemoteBytes(blockId)
-                .get))
-            val start = pid * taskSize + math.min(pid, extraSize)
-            val length = taskSize + (if (pid < extraSize) 1 else 0)
-            require(localBuffer.array().length == length * 2)
-            SerializerInstance.serialize(localBuffer).deCompress(0, localParameter, start, length)
-            BlockManagerWrapper.unlock(blockId)
-            pid
-          } catch {
-            case t : Throwable =>
-              logger.error("Error: " + ExceptionUtils.getStackTrace(t))
-              throw t
+    val bm = SparkEnv.get.blockManager
+    val tasks = (0 until partitionNum).map { pid =>
+      syncPool.submit {
+        new Callable[Int] {
+          override def call(): Int = {
+            try {
+              val blockId = getWeightBlockId(pid)
+              val localBuffer = BlockManagerWrapper.byteBufferConvert(
+                bm.getLocalBytes(blockId).getOrElse(bm.getRemoteBytes(blockId).get))
```

Comment:
Actually the first option I listed would need to be a lazy collection like a stream for that to work. Using nested `getOrElse` works too. Is this something that is common enough to put a method in the `BlockManagerWrapper`? The Spark block manager actually has a `get` method that does exactly this, but is only defined for deserialized objects.